[TT-9798] Adding docs page: How to instrument plugins with OpenTelemetry (#3184)
TykTechnologies/tyk-docs: tyk-docs/content/plugins/otel-plugins.md

---
date: 2023-08-24T13:32:12Z
title: How to instrument plugins with OpenTelemetry
menu:
  main:
    parent: "Custom Plugins"
weight: 3
aliases:
  - "/plugins/otel-plugins"
---

> 

Reviewer: Suggest adding an introduction: 

By instrumenting your custom plugins with Tyk's OpenTelemetry library, you can gain additional insights into custom plugin behaviour like time spent and exit status. Read on to see some examples of creating span and setting attributes for your custom plugins.

> ## Prerequesites

Reviewer 2: ```suggestion
## Prerequisites
```

> - Go (at the moment it's the only supported language)

Reviewer 2: Is there a specific version of Go required, e.g. 1.19?

```suggestion
- Go
```

> - Gateway instance with OpenTelemetry and DetailedTracing enabled:

Reviewer 2: Maybe add a brief sentence explaining what enabling detailed tracing does?

```go
"opentelemetry": {
  "enabled": true,
  "detailed_tracing":true
}
```

In order to instrument our plugins we will be using Tyk’s OpenTelemetry library implementation.
You can import it by running the following command:

```console

> $ go get github.com/TykTechnologies/opentelemetry

Author: @caroltyk we've been building and testing it using Go v1.19, so I added it as a recommended version (or higher ones). It should also work with older versions as well (1.17 and 1.18 at least)

```

## Create a new span from the request context

`trace.NewSpanFromContext()` is a function that helps you create a new span from the current request context. When called, it returns two values: a fresh context with the newly created span embedded inside it, and the span itself. This method is particularly useful for tracing the execution of a piece of code within a web request, allowing you to measure and analyze its performance over time.

> The function takes three parameters:

Reviewer 2: Maybe insert a blank line before this paragraph.

1. `Context`: This is usually the current request’s context. However, you can also derive a new context from it, complete with timeouts and cancellations, to suit your specific needs.
2. `TracerName`: This is the identifier of the tracer that will be used to create the span. If you do not provide a name, the function will default to using the `tyk` tracer.
3. `SpanName`: This parameter is used to set an initial name for the child span that is created. This name can be helpful for later identifying and referencing the span.

> ```go

Reviewer 2: Before the code example, add one or two sentences that explain what it shows.

    // We create a new span using the context from the incoming request.
    _, newSpan := trace.NewSpanFromContext(r.Context(), "", "GoPlugin_first-span")
    // Ensure that the span is properly ended when the function completes.
    defer newSpan.End()
    // Add a custom "Foo: Bar" header to the request.
    r.Header.Add("Foo", "Bar")
}
func main() {}
```

In your exporter (in this case, Jaeger) you should see something like this:

{{< img src="/img/plugins/span_from_context.png" alt="OTel Span from context" >}}

As you can see, the name we set is present: `GoPlugin_first-span` and it’s the first child of the `GoPluginMiddleware` span.

## Modifying span name and set status

The span created using `trace.NewSpanFromContext()` can be further configured after its creation. You can modify its name and set its status:

```go
func AddFooBarHeader(rw http.ResponseWriter, r *http.Request) {
	_, newSpan := trace.NewSpanFromContext(r.Context(), "", "GoPlugin_first-span")
	defer newSpan.End()

	// Set a new name for the span.
	newSpan.SetName("AddFooBarHeader Testing")

	// Set the status of the span.
	newSpan.SetStatus(trace.SPAN_STATUS_OK, "")

	r.Header.Add("Foo", "Bar")
}
```

> This updated span will then appear in the traces as `AddFooBarHeader Testing` with an **OK Status**. The second parameter receives a description, but this one is only valid for **ERROR** status.

Reviewer 2: ```suggestion
This updated span will then appear in the traces as `AddFooBarHeader Testing` with an **OK Status**.

The second parameter of the `SetStatus` method can accept a description parameter that is valid for **ERROR** statuses.
```

The available span statuses in ascending hierarchical order are:

- `SPAN_STATUS_UNSET`

- `SPAN_STATUS_ERROR`

- `SPAN_STATUS_OK`

This order is important: a span with an **OK** status cannot be overridden with an **ERROR** status. However, the reverse is possible - a span initially marked as **UNSET** or **ERROR** can later be updated to **OK**.

{{< img src="/img/plugins/span_name_and_status.png" alt="OTel Span name and status" >}}

Now we can see the new name and the `otel.status_code` tag with the **OK** status.

## Setting attributes

> You can set attributes on your spans, enriching each trace with additional, context-specific information by using the `SetAttributes()` function.

Reviewer 2: ```suggestion
The `SetAttributes()` function allows you to set attributes on your spans, enriching each trace with additional, context-specific information.
```

The following example illustrates this functionality using the OpenTelemetry library's implementation by Tyk

```go
func AddFooBarHeader(rw http.ResponseWriter, r *http.Request) {
	_, newSpan := trace.NewSpanFromContext(r.Context(), "", "GoPlugin_first-span")
	defer newSpan.End()

    // Set an attribute on the span.
	newSpan.SetAttributes(trace.NewAttribute("go_plugin", "1"))

	r.Header.Add("Foo", "Bar")
}
```

In the above code snippet, we set an attribute `go_plugin` with a value of `1` on the span. This is just a demonstration; in practice, you might want to set attributes that carry meaningful data relevant to your tracing needs.

Attributes are key-value pairs. The value isn't restricted to string data types; it can be any value, including numerical, boolean, or even complex data types, depending on your requirements. This provides flexibility and allows you to include rich, structured data within your spans.

> How the `go_plugin` attribute looks in Jaeger:

Reviewer 2: ```suggestion
The illustration below, shows how the `go_plugin` attribute looks in Jaeger:
```

	ctx, newSpan := trace.NewSpanFromContext(r.Context(), "", "GoPlugin_first-span")
	defer newSpan.End()

	// Set an attribute on this span.
	newSpan.SetAttributes(trace.NewAttribute("go_plugin", "1"))

	// Call another function, passing in the context (which includes the new span).
	NewFunc(ctx)

	// Add a custom "Foo: Bar" header to the request.
	r.Header.Add("Foo", "Bar")
}

func NewFunc(ctx context.Context) {
	// Start a new span for this function, using the context passed from the calling function.
	_, newSpan := trace.NewSpanFromContext(ctx, "", "GoPlugin_second-span")
	defer newSpan.End()

	// Simulate some processing time.
	time.Sleep(1 * time.Second)

	// Set an attribute on this span.
	newSpan.SetAttributes(trace.NewAttribute("go_plugin", "2"))
}
```

In this example, the `AddFooBarHeader` function creates a span and then calls `NewFunc`, passing the updated context. The `NewFunc` function starts a new span of its own, linked to the original through the context. It also simulates some processing time by sleeping for 1 second, then sets a new attribute on the second span. In a real-world scenario, the `NewFunc` would contain actual code logic to be executed.

> How this new child span looks in Jaeger:

Reviewer 2: ```suggestion
The illustration below, shows how this new child looks in Jaeger:
```

{{< img src="/img/plugins/multiple_spans.png" alt="OTel Span attributes" >}}

## Error handling

> In OpenTelemetry, it's essential to understand the distinction between recording an error and setting the span status to error. The `RecordError()` function records an error as an exception span event. However, this alone doesn't change the span's status to error. To mark the span as error, you need to make an additional call to the `SetStatus()` function.

Reviewer 2: How is the exception span event raised? Does it raise an error like in a try...catch block or is it published to subscribers/listeners?

Author: @dcs3spp you can use a try...catch block in your code (actually, the Go's equivalent) and if there's an exception, you can record the error and stop the execution of the plugin. Then, in your span, you'll see the error description and some more useful information (take a look at the last image)

---
date: 2023-08-24T13:32:12Z
title: How to instrument plugins with OpenTelemetry
menu:
  main:
    parent: "Custom Plugins"
weight: 3
aliases:
  - "/plugins/otel-plugins"
---

> By instrumenting your custom plugins with Tyk's OpenTelemetry library, you can gain additional insights into custom plugin behaviour like time spent and exit status. Read on to see some examples of creating span and setting attributes for your custom plugins.

Reviewer 3: @mativm02 isn't the time spent information already there without needing extra instrumentation? The user will be able to see the span for the go plugin independently if they add more info or not.

Author: @tbuchaillot that's correct, but you can see it in a more granular way if you add multiple spans (you can know how much time each span/function took to finish) - let me make a small modification

## Prerequisites

- Go v1.19 or higher
- Gateway instance with OpenTelemetry and DetailedTracing enabled:

```go
"opentelemetry": {
  "enabled": true,

>   "detailed_tracing":true

Reviewer 3: this configuration doesn't exist here - it's at API level

Author: @tbuchaillot but this is a prerequisite:

> Gateway instance with OpenTelemetry and DetailedTracing enabled

I think it's worth it to keep it there just to let the user know what we're talking about in that sentence

---
date: 2023-08-24T13:32:12Z
title: How to instrument plugins with OpenTelemetry
menu:
  main:
    parent: "Custom Plugins"
weight: 3
aliases:

>   - "/plugins/otel-plugins"

Reviewer 4: @dcs3spp 
1. Why do we need this alias?
2. The file is not in the correct file path

Author: @letzya 

1. I copied it from the other plugin files from the same path (`how-to-serve-plugins.md`, `supported-languages.md`, etc.) but didn't find any explanation of this in the [Contributor Guide](https://github.com/TykTechnologies/tyk-docs/blob/master/CONTRIBUTING-TECHNICAL-GUIDE.md) - should we remove this?
2. I put this file where all the plugin documentation is located. What would be the correct file path? 🤔

Reviewer 2: Hi @mativm02 Aliases are used when a page is moved to a new file location. There is a note about it in this section of the [contributors guide](https://tyktech.atlassian.net/wiki/spaces/EN/pages/1888976900/The+Docs+Bible+-+A+Guide+For+The+Docs+Contributor#New-Website-Content-Guidelines%5BhardBreak%5D). They allow a file to be accessed by different URLs (aliases). In the above case the alias can be removed in lines 8-9.

---

By instrumenting your custom plugins with Tyk's OpenTelemetry library, you can gain additional insights into custom plugin behaviour like a more granular way of viewing the time spent by each function and the exit status. Read on to see some examples of creating span and setting attributes for your custom plugins.

## Prerequisites

- Go v1.19 or higher

> - Gateway instance with OpenTelemetry and DetailedTracing enabled

Author: @SonjaChevre thanks! We'll wait for Sredny's documentation to be merged so we can add this.

> At the API level add this configuration:

Reviewer 3: I'd specify it's in the API definition instead of "At the API level".  What do you think @dcs3spp ?

```json
{
	"detailed_tracing": true
}
```

and in the Gateway config:
```json
{
	"opentelemetry": {
		"enabled": true
	}
}
```

> **Note**: DetailedTracing is required to see the plugin spans in the traces.

Reviewer 2: Update this so it appears in a note bounded box
